Refactor hasher lambda, define hash store interface, mypy

schatten opened the pull request:
Summary
---------
The PR turned out to be larger than I had imagined. It does a lot.
1. Adds mypy as a dependency and begins to add type hints to the lambdas. Adds some typing for boto using `boto3-stubs`
1. Converts arbitrary wire-messages (SQS, Dynamo) for hasher lambda into formal data structures
1. Defines a simple wrapper (HashStore) over dynamodb for maintaining hash records, expect more methods from the CLI/Admin surface to go into the wrapper
1. docstring for hasher lambda

Test Plan
---------
```
$> mypy
$> make upload_docker
$> edit hma_lambda_docker_uri variable to point to new sha256 in ECR
$> terraform apply -var "prefix=$(whoami)"
```

Followed by uploading an image to S3 and receiving a match on email.

Note: This was also the first time I edited an existing terraform instance. And it worked out well. The lambdas were switched out and the architecture picked things up automatically.

schatten commented:
As can be divined from the branch name, i attempted to add unittests, but couldn't resist the urge to deliver this refactor.

Dcallies commented on hasher-matcher-actioner/hmalib/dto.py:
```diff
@@ -0,0 +1,34 @@
+import datetime
+import typing as t
+
```
Nit: Docstring plox

Dcallies commented on hasher-matcher-actioner/hmalib/dto.py:
```diff
@@ -0,0 +1,34 @@
+
+
+class PDQHashRecord(t.NamedTuple):
+    """Successful execution at the hasher produces this record.
```
Nit: 

```
"""
Comment goes here
"""
```

Dcallies commented on hasher-matcher-actioner/hmalib/dto.py:
```diff
@@ -0,0 +1,34 @@
+    quality: int
+    timestamp: datetime.datetime  # ISO-8601 formatted
+
+    def to_dynamodb_item(self) -> dict:
```
Don't use dict, use t.Dict!

mypy strict will also force to type it-> t.Dict[str, ...]

Dcallies commented on hasher-matcher-actioner/hmalib/dto.py:
```diff
@@ -0,0 +1,34 @@
+
+    def to_dynamodb_item(self) -> dict:
+        return {
+            "PK": "c#{}".format(self.content_key),
```
Use format strings

f"c#{self.content_key}"

Dcallies commented on hasher-matcher-actioner/hmalib/lambdas/pdq/pdq_hasher.py:
```diff
@@ -9,60 +9,77 @@
+                    MessageBody=json.dumps(hash_record.to_sqs_message()),
                 )
+
+                logger.info("Published new PDQ hash")
```
Should this print the PDQ hash? Seems useful, or maybe the key?

Dcallies commented on hasher-matcher-actioner/hmalib/storage/hashstore.py:
```diff
@@ -0,0 +1,16 @@
+from mypy_boto3_dynamodb.service_resource import Table
+
+class _IDynamoDBItem():
+    """ Clowny type as an interfaces. True interfaces in python are nightmares.
```
Nit: inconsistent formatting

Dcallies commented on hasher-matcher-actioner/hmalib/storage/hashstore.py:
```diff
@@ -0,0 +1,16 @@
+    def to_dynamodb_item(self) -> dict:
+        pass
+
+class HashStore:
```
blocking: I don't think this is a HashStore. I think this is the MatchStore, MatchRecords, PDQMatches, or some variation.

@BarrettOlson , thoughts?

Dcallies commented on hasher-matcher-actioner/mypy.ini:
```diff
@@ -0,0 +1,2 @@
+[mypy-threatexchange.hashing.*]
+ignore_missing_imports = True
```
lulz

Dcallies reviewed (approved):
LGTM

schatten commented:
Did another round of testing using the test plan after addressing nits and a small change to make it easier to get the PK out of the DTO.

It works!

BarrettOlson commented on hasher-matcher-actioner/hmalib/dto.py:
```diff
@@ -0,0 +1,39 @@
+    def to_dynamodb_item(self) -> dict:
+        return {
+            "PK": PDQHashRecord.get_dynamodb_pk(self.content_key),
+            "SK": "type:pdq",
```
@schatten was the change from 'type#pdq' to 'type:pdq' intentional? Going to change it back in my next PR if not. Thanks!

schatten replied:
Nope. Totally unintentional. Nice spot. Thanks!
